### PR TITLE
Use same initialization in batch

### PR DESCRIPTION
**Pull request by tarokiritani**
In SPSA, randomization was done for each input in a batch. This made the
implementation batch inconsistent.

### CODE REVIEW

**Comment by feinsteinben** [comment]
@tarokiritani I think that there is a mix of several changes here.
1. Changing the loss - you change works when the output shape is B×1. I think that supporting the more general case is better, but it's also ok to limit ourselves and document the requirements.
2. Making this algorithm batch consistent (and require less random numbers) - this change is welcome.
3. Copying the data at every iteration (contiguous), and allocating space for `vb` at every iteration. I think this change will be expensive in memory consumption and run time.

File: advertorch/attacks/spsa.py [hunk 1]
```diff
@@ -107,39 +108,23 @@ def spsa_perturb(predict, loss_fn, x, y, eps, delta, lr, nb_iter,
...
+            vb = torch.empty_like(xb[:, 0:1, ...])
             vb = vb.bernoulli_().mul_(2.0).sub_(1.0)
-            if ii == nb_batch - 1 and nb_batch * batch_size > nb_sample:
-                x_ = xb[:nb_batch * batch_size - nb_sample]
-                y_ = yb[:nb_batch * batch_size - nb_sample]
-                v_ = vb[:nb_batch * batch_size - nb_sample]
-            else:
-                x_ = xb
-                y_ = yb
-                v_ = vb
```
Comment (feinsteinben) [inline advertorch/attacks/spsa.py]:
```python
vb = vb.bernoulli_().mul_(2.0).sub_(1.0)
x_ = xb[:len(nb_sample_per_batch)]
y_ = yb[:len(nb_sample_per_batch)]
v_ = vb[:len(nb_sample_per_batch)]
```

File: advertorch/attacks/spsa.py [hunk 2]
```diff
@@ -107,39 +108,23 @@ def spsa_perturb(predict, loss_fn, x, y, eps, delta, lr, nb_iter,
...
-
     x = x.unsqueeze(0)
     y = y.unsqueeze(0)
-    dx = torch.zeros_like(x)
+    dx = torch.zeros_like(x) # is it necessary to define x after the expansin of x?
     dx.grad = torch.zeros_like(dx)
     optimizer = torch.optim.Adam([dx], lr=lr)
-    xb = x.expand(batch_size, *x.shape[1:]).contiguous()
-    yb = y.expand(batch_size, *y.shape[1:]).contiguous()
-    vb = torch.empty_like(xb)
```
Comment (feinsteinben) [inline advertorch/attacks/spsa.py]:
```python
    xb = x.expand(max_batch_size, *x.shape[1:]).contiguous()
    yb = y.expand(max_batch_size, *y.shape[1:]).contiguous()
    vb = torch.empty((max_batch_size, 1, *x.shape[2:]), dtype=x.dtype, device=x.device)
```

File: advertorch/attacks/spsa.py [hunk 3]
```diff
@@ -107,39 +108,23 @@ def spsa_perturb(predict, loss_fn, x, y, eps, delta, lr, nb_iter,
...
-            else:
-                x_ = xb
-                y_ = yb
-                v_ = vb
-
-            grad = spsa_grad(predict, loss_fn, x_ + dx, y_,
-                             v_, delta, reduction="sum")
+            vb = vb.expand_as(xb).contiguous()
+            grad = spsa_grad(predict, loss_fn, xb + dx, yb,
+                    vb, delta, reduction="sum")
```
Comment (feinsteinben) [inline advertorch/attacks/spsa.py]:
```python
vb = v_.expand_as(x_).contiguous()
grad = spsa_grad(predict, loss_fn, x_ + dx, y_,
                 v_, delta, reduction="sum")
```

**Comment by tarokiritani** [comment]
Your suggestion makes sense, I suppose. Let me work on it.

**Comment by tarokiritani** [comment]
I addressed your points. The test will pass with #4 .

File: advertorch/attacks/spsa.py [hunk 4]
```diff
@@ -107,39 +108,25 @@ def spsa_perturb(predict, loss_fn, x, y, eps, delta, lr, nb_iter,
...
-                y_ = yb[:nb_batch * batch_size - nb_sample]
-                v_ = vb[:nb_batch * batch_size - nb_sample]
-            else:
-                x_ = xb
-                y_ = yb
-                v_ = vb
-
+            x_ = xb[:len(nb_sample_per_batch)]
+            y_ = yb[:len(nb_sample_per_batch)]
+            v_ = vb.expand_as(x_).contiguous()
```
Comment (feinsteinben) [inline advertorch/attacks/spsa.py]: I think that it should be
```python
v_ = vb[:len(nb_sample_per_batch)].expand_as(x_).contiguous()
```
otherwise it may fail when the last batch is smaller.

Comment (tarokiritani) [reply advertorch/attacks/spsa.py]: Good catch. Will be fixed in a minute.

Comment (tarokiritani) [reply advertorch/attacks/spsa.py]: This is addressed in c2bc435.

File: advertorch/attacks/spsa.py (hunk 3 above)
Comment (tarokiritani) [reply advertorch/attacks/spsa.py]: Resolved in 176d7c6.

File: advertorch/attacks/spsa.py (hunk 2 above)
Comment (tarokiritani) [reply advertorch/attacks/spsa.py]: Resolved in 176d7c6.

File: advertorch/attacks/spsa.py (hunk 1 above)
Comment (tarokiritani) [reply advertorch/attacks/spsa.py]: Resolved in 176d7c6.

**Comment by feinsteinben** [comment]
@tarokiritani lgtm
